Keep `results_count` specific to each char name

lekoenig opened the pull request:
This PR includes code changes to 1) sum `results_count` by unique characteristic names when building `p2_site_counts`, and 2) modify `add_download_groups()` so that separate task numbers (and download groups) get assigned for unique characteristics. These code changes are meant to capture the intended behavior in the current pipeline, but get rid of the empty data calls that result from using `cross` to map over all characteristic names when building `p2_wqp_data_aoi`. The resulting data should not be affected by these changes. 

Requesting a review from @lindsayplatt because these changes may impact the decisions made in the national chlorophyll pipeline. Also tagging @jordansread for awareness. 

Closes #88

lindsayplatt commented on 2_download/src/fetch_wqp_data.R:
```diff
@@ -96,10 +96,13 @@ add_download_groups <- function(sitecounts_df, max_sites = 500, max_results = 25
+        mutate(task_num_by_results = MESS::cumsumbinning(x = results_count, 
+                                                         threshold = max_results, 
+                                                         maxgroupsize = max_sites), 
+               task_num = task_num_by_results - 1 + cur_group_id()) %>%
```
This is very clever and I had to go in and add `browser()` and compare `task_num_by_results` to `task_num` after this ran to understand it. Maybe a short little comment here? Especially because the comments above this don't mention that groups are then split based on `CharacteristicName`!

lekoenig replied:
Yeah, good idea! Do you think the text in https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/92/commits/11fa6b342db5a2579d08b2f65d56659798834ea1 helps here?

lindsayplatt replied:
Yes, but I did have to re-read a couple of times to fully understand. Maybe a small example would help?

> For example, both "Specific conductance" and "Temperature" may have values for `task_num_by_results` of 1 and 2 but the values of `cur_group_id()` (1 and 2, respectively) would mean that they have unique values for `task_num` equaling 1, 2, 3, and 4.

lindsayplatt reviewed (approved):
I love this solution to not query characteristic names with no results. I thought it was going to be much more complicated to code that up, but wow!

lekoenig commented:
> I thought it was going to be much more complicated to code that up

@lindsayplatt, your hunch may have been right 😅. I tried another version of my reprex when adding the documentation and I think I may have just gotten lucky that this simple solution worked for a particular example where there were only 1 or 2 `task_num`'s. I may need to revisit this code chunk (so sorry!), see https://github.com/USGS-R/ds-pipelines-targets-example-wqp/issues/88#issuecomment-1222891785.

lekoenig commented:
In https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/92/commits/31203f89e7d45d17eb148827bd666e7d47de491a I've implemented a change to how `task_num` is assigned for each group of sites. 

The original code in this PR would probably work much of the time because our threshold for `results_max` is fairly high (250,000 records). Here's an example reprex that shows an edge case, though:
```r
sitecounts <- tibble(MonitoringLocationIdentifier = c("A","B","C","D","E","F","A","A"), 
                    lon = c(rep(-92.4, 8)), 
                    lat = c(rep(42.5,8)), 
                    datum = c(rep("WGS84",8)), 
                    grid_id = c(rep("11923", 8)), 
                    CharacteristicName = c(rep("Temperature, water", 3), rep("Temperature", 3),"Nitrate","conductivity"), 
                    results_count = c(151, 247000, 7500,15000,235004,134, 249000, 1100))
```

My original solution in this PR doesn't actually work as intended because `task_num == 4` applies to two separate characteristic names, `"Temperature"` and `"Temperature, water"`:

```r
> sitecounts %>%
+    rename(site_id = MonitoringLocationIdentifier) %>% 
+    split(.$grid_id) %>%
+    purrr::map_dfr(.f = function(df){
+      
+        df_grouped <- df %>%
+            group_by(CharacteristicName) %>%
+            arrange(desc(results_count)) %>%
+            mutate(task_num_by_results = MESS::cumsumbinning(x = results_count, 
+                                                             threshold = max_results, 
+                                                             maxgroupsize = max_sites), 
+                   task_num = task_num_by_results - 1 + cur_group_id()) %>%
+            ungroup()
+    })
# A tibble: 8 x 9
  site_id   lon   lat datum grid_id CharacteristicName results_count task_num_by_results task_num
  <chr>   <dbl> <dbl> <chr> <chr>   <chr>                      <dbl>               <int>    <dbl>
1 A       -92.4  42.5 WGS84 11923   Nitrate                   249000                   1        2
2 B       -92.4  42.5 WGS84 11923   Temperature, water        247000                   1        4
3 E       -92.4  42.5 WGS84 11923   Temperature               235004                   1        3
4 D       -92.4  42.5 WGS84 11923   Temperature                15000                   2        4
5 C       -92.4  42.5 WGS84 11923   Temperature, water          7500                   2        5
6 A       -92.4  42.5 WGS84 11923   conductivity                1100                   1        1
7 A       -92.4  42.5 WGS84 11923   Temperature, water           151                   2        5
8 F       -92.4  42.5 WGS84 11923   Temperature                  134                   2        4
>
```

However, using the code in https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/92/commits/31203f89e7d45d17eb148827bd666e7d47de491a, we get unique `task_num` (and `download_grp`) for each characteristic name:
```r
> add_download_groups(sitecounts)
# A tibble: 8 x 9
  site_id   lat   lon datum grid_id CharacteristicName results_count download_grp pull_by_id
  <chr>   <dbl> <dbl> <chr> <chr>   <chr>                      <dbl> <chr>        <lgl>     
1 A        42.5 -92.4 WGS84 11923   conductivity                1100 11923_1      TRUE      
2 A        42.5 -92.4 WGS84 11923   Nitrate                   249000 11923_2      TRUE      
3 E        42.5 -92.4 WGS84 11923   Temperature               235004 11923_3      TRUE      
4 D        42.5 -92.4 WGS84 11923   Temperature                15000 11923_4      TRUE      
5 F        42.5 -92.4 WGS84 11923   Temperature                  134 11923_4      TRUE      
6 B        42.5 -92.4 WGS84 11923   Temperature, water        247000 11923_5      TRUE      
7 C        42.5 -92.4 WGS84 11923   Temperature, water          7500 11923_6      TRUE      
8 A        42.5 -92.4 WGS84 11923   Temperature, water           151 11923_6      TRUE      
>
```

lekoenig commented on 2_download/src/fetch_wqp_data.R:
```diff
@@ -83,23 +83,49 @@ add_download_groups <- function(sitecounts_df, max_sites = 500, max_results = 25
+            filter(char_group == df_grouped$char_group[i] - 1) %>%
+            pull(task_num) %>%
+            max()
+          df_grouped$task_num[i] <- df_grouped$task_num_by_results[i] + max_task_prev_group
```
>  Is there any reason that the download groups need to be numeric only?

Not really, the primary reason I didn't concatenate the characteristics with the task numbers was just to print a more interpretable message to the console when downloading the data (i.e., "Fetching data for 11923_1, Specific conductance").

lindsayplatt replied:
Good point. I actually just thought about one other method that may be simple but worth thinking about quickly. Ultimately, the unique combinations of `char_group` and `task_num_by_results` are what we want the task numbers to be. What if we just did a new `group_by()` and then made the `task_num` the `cur_group_id()`? I get the same result doing that:

```r
library(tidyverse)

max_results <- 250000
max_sites <- 500

sitecounts <- tibble(MonitoringLocationIdentifier = c("A","B","C","D","E","F","A","A"), 
                     lon = c(rep(-92.4, 8)), 
                     lat = c(rep(42.5,8)), 
                     datum = c(rep("WGS84",8)), 
                     grid_id = c(rep("11923", 8)), 
                     CharacteristicName = c(rep("Temperature, water", 3), rep("Temperature", 3),"Nitrate","conductivity"), 
                     results_count = c(151, 247000, 7500,15000,235004,134, 249000, 1100))

# This is the code you have in your PR
x <- sitecounts %>%
  rename(site_id = MonitoringLocationIdentifier) %>% 
  split(.$grid_id) %>%
  purrr::map_dfr(.f = function(df){
    
    df_grouped <- df %>%
      group_by(CharacteristicName) %>%
      arrange(desc(results_count)) %>%
      mutate(task_num_by_results = MESS::cumsumbinning(x = results_count, 
                                                       threshold = max_results, 
                                                       maxgroupsize = max_sites), 
             char_group = cur_group_id(),
             task_num = NA_integer_) %>%
      ungroup()
  }) %>% arrange(char_group, task_num_by_results)

x %>% 
  group_by(char_group, task_num_by_results) %>% 
  mutate(task_num = cur_group_id()) %>% 
  ungroup()

# A tibble: 8 x 10
# Groups:   char_group, task_num_by_results [6]
  site_id   lon   lat datum grid_id CharacteristicName results_count task_num_by_results char_group task_num
  <chr>   <dbl> <dbl> <chr> <chr>   <chr>                      <dbl>               <int>      <int>    <int>
1 A       -92.4  42.5 WGS84 11923   conductivity                1100                   1          1        1
2 A       -92.4  42.5 WGS84 11923   Nitrate                   249000                   1          2        2
3 E       -92.4  42.5 WGS84 11923   Temperature               235004                   1          3        3
4 D       -92.4  42.5 WGS84 11923   Temperature                15000                   2          3        4
5 F       -92.4  42.5 WGS84 11923   Temperature                  134                   2          3        4
6 B       -92.4  42.5 WGS84 11923   Temperature, water        247000                   1          4        5
7 C       -92.4  42.5 WGS84 11923   Temperature, water          7500                   2          4        6
8 A       -92.4  42.5 WGS84 11923   Temperature, water           151                   2          4        6
```

lekoenig replied:
Ooh, I liked your first suggestion for replacing the for loop, but I think I like this one even better because it's still simple but yet we don't have to mess with the `sitecounts_bad_ids` naming scheme! I've implemented these changes in https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/92/commits/e2130c091c470e4d5129cee6f1e15570de078f24.

lindsayplatt reviewed (approved):
Looks good! Nice catch on that more tricky edge case. Glad we solved it!

lekoenig commented:
> Looks good! Nice catch on that more tricky edge case. Glad we solved it!

Thanks for your suggestions and for reviewing this (again) after I found that unintended behavior. Glad we figured it out!
